proposals/Exceptions.md

```diff
@@ -38,125 +39,136 @@ and the instructions *throw* and *rethrow*.
+constructor must be on top of the value stack, and must correspond to the
+exception type signature for the exception.
 
-Otherwise, the call stack is wound back to the function with the nearest
-enclosing try block. The value stack is then popped back to the size the value
-stack had when that try block was initially entered.
+When an exception is thrown, the values on the stack (corresponding to the type
+signature) are popped off and the corresponding exception is constructed.  The
+exception is stored internally for access when the exception is caught. The
+runtime then searches for nearest enclosing try block body that execution is
+in. That try block is called the _catching_ try block.
 
-If the throw appears within a catch block, it is treated as being outside the
-corresponding try block of the catch (i.e. just after the end of corresponding
-try block). Therefore, throws within the catch block can't be caught by their
-enclosing try block. Rather, execution is thrown to the next enclosing try
+If the throw appears within the body of a try block, it is the catching try
 block.
 
-Once the value stack has been popped back to the size of the enclosing try block
-(whether local or across function calls), control is then passed to the
-appropriate catch block.
-
-If there is a typed catch block with the same exception type _index_ as the
-throw, the thrown values (stored internally at the throw) are copied back on the
-value stack. Then, the corresponding typed catch block is entered. This allows
-the values of the exception to be accessed by the catch block.
-
-On the other hand, if there isn't a corresponding typed catch block, and there
-is a default catch block, the default catch block is entered without pushing the
-values of the exception onto the value stack. This allows the default catch
-block to clean up state, but it can't look at the thrown value.
+If a throw occurs within a function body, and it doesn't appear inside the body
+of a try block, the throw continues up the call stack until it is in the body of
+an an enclosing try block, or the call stack is flushed. If the call stack is
+flushed, execution is terminated. Otherwise, the found enclosing try block is
+the catching try block.
+
+A throw inside a the body of a catch block is never caught by the corresponding
+try block of the catch block, since instructions in the body of the catch block
+are not in the body of the try block.
+
+Once a catching try block is found for the throw, the value stack is popped back
+to the size the value stack had when the try block was entered. Then, tagged
+catch blocks are tried in the order they appear in the catching try block, until
+one matches. A tagged catch block matches if the corresponding `catch`
+instruction is applicable, i.e. the corresponding exception was generated using
+the constructor defined by the tag of the `catch` instruction.
+
+If a matched tagged catch block is found, control is transferred to the body of
```

Comment:
Replaced this sentence with "If a matched tagged catch block is found, control is transferred to the body of the catch block, and the data fields of the exception are pushed back onto the stack."